CLN: move plotting funcs to pd.plotting

sinhrks opened the pull request:
Rebased in #16005

---

- [x] closes #12548
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

cleanup `tools/plotting` as `plotting` subpackage. Because the differences are being huge, would like to split 3 part if base direction is OK:

~~1. Move plotting related tests under `pandas/tests/plotting` splitting to `Series`, `DataFrame`, etc. (0.18.2)~~ (#13621)
2. Create `pd.plotting` and move related files (0.20.0) 
3. Split / refactor `pd.plotting` to use small utility functions / property control classes. Add tests. (0.20.0)

sinhrks committed:
CLN: move plotting funcs to pd.plotting

jorisvandenbossche commented:
@sinhrks Generally sounds like a good plan!

TomAugspurger commented:
+1 overall. @sinhrks will this break old code? Does `pandas.tools.plotting` just point to `pandas.plotting`? I looked in the diff, but could have missed it.

TomAugspurger commented on pandas/plotting/compat.py:
```diff
@@ -0,0 +1,51 @@
+    try:
+        import matplotlib as mpl
+        return (str(mpl.__version__) <= LooseVersion('1.2.1') and
+                str(mpl.__version__)[0] != '0')
```
I had to add these checks for `'0'` when matplotlib's versioneer was broken for 1.5. I believe it's fixed now. Do you want to remove them in here, or a separate PR?

jorisvandenbossche commented:
@sinhrks Can you rebase this? Further it is mainly docs that is still needed?

jreback commented:
@sinhrks as usual, certainly can include in 0.19.0, just trying to clear the decks.

jreback commented:
@sinhrks can you update

jorisvandenbossche commented:
@sinhrks do you time to update this?

jorisvandenbossche commented:
@sinhrks Do you have time to update this? I still think this is a useful refactor. I can also do a rebase to get it in.

sinhrks commented:
Once updated to the latest master. Let me add release note later.

jreback commented on pandas/tseries/plotting.py:
```diff
@@ -1,313 +1,3 @@
-    subplot.format_coord = lambda t, y: (
-        "t = {0}  y = {1:8f}".format(Period(ordinal=int(t), freq=freq), y))
-
-    pylab.draw_if_interactive()
```
is there a reason you left this file? (for some sort of compat)?

jreback replied:
or I see, you may have left all of the pandas/tseries namespace as is (and just moved things), ok then.

jorisvandenbossche commented on pandas/plotting/api.py:
```diff
@@ -0,0 +1,20 @@
+"""
+Plotting api
```
I know this is a bit the style in other pandas modules as well, but can't we just put this in `__init__.py` ?

jreback replied:
I think this is cleaner here actually. And it is only imported explicity (unlike ``__init__`` which is implicity imported)

jorisvandenbossche replied:
What do you mean with implicitly imported?

The difference with the other modules where we use this pattern, is that, eg in `pandas.io`, `pandas.io.api` are those functions that are imported top-level (imported in `pandas/__init__.py`), not what makes up the `pandas.io` namespace itself. Here, we don't have a collection of functions (put in `api.py`) which we want to have top level, but we just have the `pandas.io` namespace. So IMO it makes more sense and is cleaner to have it just in the init file here.

jreback replied:
what I mean is, unless you explicity reference these imports, like ``from pandas.plotting import .....)``  you would by-default actually import anything.

thoughI guess for back-compat this is needed.

IOW, the imports for things like ``matplotlib`` would *only* be done when the pandas.plotting namespace is used (IOW, you try to plot something), so the import is done lazily later, rather than on startup now (to register converters and such).

jorisvandenbossche commented on pandas/plotting/api.py:
```diff
@@ -0,0 +1,20 @@
+# flake8: noqa
+
+try:  # mpl optional
+    from pandas.plotting import converter as conv
```
I would leave out the `conv` (now both `conv` and `converter` in pandas.plotting namespace)

jorisvandenbossche commented on pandas/plotting/api.py:
```diff
@@ -0,0 +1,20 @@
+                                  andrews_curves, bootstrap_plot,
+                                  parallel_coordinates, lag_plot,
+                                  autocorrelation_plot)
+from pandas.plotting.plotting import (boxplot, scatter_plot, grouped_hist,
```
Not the most important :-), but the `plotting.plotting` looks a bit silly. Maybe we can call this file `base` or `core`  to make it more clear?

I am also not sure if `grouped_hist`, `hist_frame`, `hist_series` should be exposed in the public `pandas.plotting` namespace?

jorisvandenbossche reviewed (commented):
Looks good, thanks!

And sorry, I merged a PR on the converters. I had better waited to merge that one until this is merged, as you now have to rebase again (for the converters.py file)

Some other things:
- I would personally move `tests/plotting` also into this new module (so `plotting/tests`)
- Should we raise a deprecation warning on the old namespaces? So we can remove them sometime.

sinhrks committed:
Add deprecation wrapper

jorisvandenbossche commented on pandas/tools/plotting.py:
```diff
@@ -1,3 +1,21 @@
+    def outer(t=t):
+
+        def wrapper(*args, **kwargs):
+            warnings.warn("pandas.tools.plotting{t} is deprecated. "
```
`plotting{t}` -> plotting.{t}

jorisvandenbossche commented on pandas/tools/plotting.py:
```diff
@@ -1,3 +1,21 @@
+            warnings.warn("pandas.tools.plotting{t} is deprecated. "
+                          "import from the "
+                          "pandas.plotting.{t} instead".format(t=t),
+                          UserWarning, stacklevel=3)
```
Can you make this a FutureWarning instead of UserWarning (I think FutureWarning is OK here, as most of those functions will be used by users directly)

jorisvandenbossche commented on pandas/tools/plotting.py:
```diff
@@ -1,3 +1,21 @@
+        def wrapper(*args, **kwargs):
+            warnings.warn("pandas.tools.plotting{t} is deprecated. "
+                          "import from the "
+                          "pandas.plotting.{t} instead".format(t=t),
```
I think the "from the" can be left out.

I would do either: 
"pandas.tools.plotting.scatter_matrix is deprecated, import pandas.plotting.scatter_matrix instead" 
or 
"pandas.tools.plotting.scatter_matrix is deprecated, import from pandas.plotting instead"

jorisvandenbossche reviewed (commented):
some comments on the deprecation.

Wondering, should we deprecate the top-level `pd.scatter_matrix` as well? (pointing to `pd.plotting.scatter_matrix`)

sinhrks committed:
Move tests upder plotting and other corrections

sinhrks committed:
Do not import deprecate decorator on the top namespace

sinhrks committed:
addd whatsnew

sinhrks commented:
Thx, updated based on the comments.

- tests are moved to ``pandas/plotting/tests``.
- ``pd.scatter_matrix`` has been deprecated, in favor of ``pd.plotting.scatter_matrix``.
- added tests to check warnings are properly raised.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -194,6 +195,31 @@ Using ``.iloc``. Here we will get the location of the 'A' column, then use *posi
+top namespace ``pandas.plotting``. All the public plotting functions should be available
+from ``pandas.plotting``.
+
+Also, ``scatter_matrix`` function imported directly under ``pandas`` namespace is also deprecated.
```
``pandas.scatter_matrix``

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -194,6 +195,31 @@ Using ``.iloc``. Here we will get the location of the 'A' column, then use *posi
+Deprecate .plotting
+^^^^^^^^^^^^^^^^^^^
+
+``pandas.tools.plotting`` module has been deprecated, moving directory under the
```
The ``pandas.tools.plotting`` module has been deprecated, in favor of the top level namespace ``pandas.plotting``

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -194,6 +195,31 @@ Using ``.iloc``. Here we will get the location of the 'A' column, then use *posi
+Previous script:
+
+.. code-block:: python
+
```
add ``import pandas as pd``

jreback commented on pandas/__init__.py:
```diff
@@ -46,7 +46,15 @@
-from pandas.tools.plotting import scatter_matrix, plot_params
+
+# deprecate tools.plotting, and scatter_matrix on the top namespace
+import pandas.tools.plotting
```
you can actually deprecate ``pandas.tools.plotting`` with pandas.util.deprecate_module

jreback commented on pandas/__init__.py:
```diff
@@ -46,7 +46,15 @@
+import pandas.tools.plotting
+from pandas.plotting import plot_params
+# do not import deprecate to top namespace
+scatter_matrix = pandas.util.decorators.deprecate(
```
any reason NOT to remove ``plot_parms`` as well from top-level?

jorisvandenbossche replied:
+1 on moving plot_params as well

jreback commented:
really just some minor comments. ok with merging, then fixing after if it is easier (one test failure on locale on travis, not sure why that is).

jorisvandenbossche commented on pandas/plotting/api.py:
```diff
@@ -0,0 +1,20 @@
+                                  parallel_coordinates, lag_plot,
+                                  autocorrelation_plot)
+from pandas.plotting.core import (boxplot, scatter_plot, grouped_hist,
+                                  hist_frame, hist_series)
```
I think I mentioned this before, but is there use in exposing `grouped_hist`, `hist_frame`, `hist_series` as public functions? 
There are also not documented I think

jorisvandenbossche replied:
The documentation argument is actually also true for `boxplot` and `scatter_plot`, and for both the functionality is also available in DataFrame.plot.scatter or DataFrame.boxplot/plot.box

jreback commented:
@sinhrks seems really close...can you rebase

jorisvandenbossche commented:
I rebased this one, but pushed to a new PR: https://github.com/pandas-dev/pandas/pull/16005

(this PR predates from when you can push to PR's branch)
